larray_editor/arrayadapter.py

```diff
@@ -3,188 +3,600 @@
 import numpy as np
 import larray as la
 
-from larray_editor.utils import Product, _LazyDimLabels
+from larray_editor.utils import Product, _LazyDimLabels, Axis, get_sample
 
 
-class LArrayDataAdapter(object):
-    def __init__(self, axes_model, xlabels_model, ylabels_model, data_model):
-        # set models
+REGISTERED_ADAPTERS = {}
+
+def register_adapter(type):
+    """Class decorator to register new adapter
+
+    Parameters
+    ----------
+    type : type
+        Type associated with adapter class.
+    """
+    def decorate_class(cls):
+        if type not in REGISTERED_ADAPTERS:
+            REGISTERED_ADAPTERS[type] = cls
+        return  cls
+    return decorate_class
+
+
+def get_adapter(data, changes, bg_value, axes_model, hlabels_model, vlabels_model, data_model):
+    data_type = type(data) if data is not None else data
+    if data_type not in REGISTERED_ADAPTERS:
+        raise ValueError("No Adapter implemented for data with type {}".format(data_type))
+    adapter_cls = REGISTERED_ADAPTERS[data_type]
+    return adapter_cls(data, changes, bg_value, axes_model, hlabels_model, vlabels_model, data_model)
+
+
+@register_adapter(None)
+class AbstractAdapter(object):
+    def __init__(self, data, changes, bg_value, axes_model, hlabels_model, vlabels_model, data_model):
```

Comment:
I also saw the Adapter as a "ModelsOrchestrator" so that it looks like the widget communicate with one model. But I agree in that case the name "Adapter" is wrong. OK to implement a true Adapter.

Reply:
We already discussed this. The goal is to make it as easy/fast as possible to write adapters for other types, so it needs to be as small as possible. Besides it is usually better to separate concerns. You could create a ModelsOrchestrator class to do that part. I do not see the value but why not, I am -0 on that, however I am a strong -1 for having the adapter doing both the orchestrator part and the adapter part (ie provide a uniform interface for different objects). The last commits you did are thus going in the right direction.